1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,248 @@
+
+This whole slow, manual approach is probably why many people forgo writing contracts this way, and opt for verbal contracts instead.It probably would not be a huge jump to say the "startup mindset" (and "agile") are one of a few potential driving factors in this "just get on with it" approach. There are many technical leaders inexperienced in this area telling their team "We'll write lovely docs later when we've got more time!" , without realizing their lack of clarity is slowing down initial development, client integrations, increasing testing efforts, and causing bizarre production issues and throughout their ecosystem.
+
+.It's pretty common to blame REST for not forcing a type system onto everyone involved in the API, but they're absolutely useable. Don't blame a paradigm for implementation level decisions.
+image::./images/contracts-unspecified-json.jpeg[A cyclist jams a stick into hisi spokes then complains about it. A pretty accurate metaphor for using unspecified JSON in a REST API then blaming REST for it.]
+
+Anyway, despite being this medium of API contracts being time consuming, it is also impossible to enforce. After you have spent the time writing up a Google Docs/HTML/Wiki, the only output of that is going to be... a Google Doc/HTML/Wiki and maybe a PDF if you want to go wild. You can't exactly jam that Google Doc into an API test suite and have it confirm that the API is conforming to whatever is written in there.
+
+There is another way.
+
+=== Description Languages
+
+A description language can be a programming language based, or text based format which lets you describe an API in a reusable format. This reusable format means you can do a whole bunch of stuff:
+
+- Documentation
+- Client-side validation
+- Server-side validation
+- Client-library Generation (SDKs)
+- UI Generation
+- Server/Application generation
+- Mock servers
+- Contract testing
+
+SOAP as been mentioned as one system that used to have meta-data which described its methods and messages using something called a WSDL:
+
+[quote, Wikipedia]
+----
+The Web Services Description Language is an XML-based interface definition language that is used for describing the functionality offered by a web service. The acronym is also used for any specific WSDL description of a web service, which provides a machine-readable description of how the service can be called, what parameters it expects, and what data structures it returns. Therefore, its purpose is roughly similar to that of a type signature in a programming language.
+----
+
+WSDLs were only used for SOAP, and not many paradigms or implementations seemed to bother with this sort of description language for a long time. Luckily that has all changed over the last few years.
+
+==== Endpoint-based APIs
```

Comment:
This might just be me not knowing things, but what do you mean by "Endpoint-based APIs"? Are there any APIs which are not endpoint-based? (genuine question, or my brain is tired and I'm missing something obvious :P). This is the first time this term is used so might be worth clarifying?